### ISSUE TITLE
MySQL override_settings only after 2nd provisioning

### ISSUE THREAD

**Issue by bitmatix** [puphpet/puphpet#2502]
Currently there seems to be the problem that the settings for mysqld doesn't get applied (to "/etc/mysql/my.cnf") for the the first provisioning. I've got to do a "vagrant halt" and then a "vagrant up --provision".

The MySQL settings in my config.yaml are:

```
mysql:
    install: '1'
    settings:
        version: '5.7'
        root_password: mysql2k16
        override_options:
            mysqld:
                sql_mode: 'STRICT_TRANS_TABLES,NO_ENGINE_SUBSTITUTION'
        restart: true
```

**Comment by pauloelr** [comment]
Same here.

I dig a little deeper and seems like the `my.cnf` file is being created before other change that replaces it.
When the file is replaced the old one is saved as `my.cnf.bak`, and this one is holding the correct values
So it should be just a question of adding a `require` instruction on the right place on puppet to solve it

**Comment by marcosgdf** [comment]
Duplicate of #1303

**Comment by tkreijne** [comment]
I''m having the same problem, It might be related to #1303 but I don't believe it's duplicate. The problem is indeed as @pauloelr already stated that on a vagrant up the correct config seems to be overwritten later in the process and thus disappears in the my.cnf.bak. After a provision it's correctly in the my.cnf so I think something is going wrong in the order in which things happen. I'm pretty new to this so I can't really get a clear overview yet.

**Comment by ethanf22** [comment]
I agree that this doesn't seem to be a duplicate of #1303. I tried doing `vagrant up`, `vagrant ssh`, `sudo service mysql restart` and the incorrect settings persisted. The settings were only corrected after doing `vagrant provision`.

**Comment by tkreijne** [comment]
Is there anyone who might have an idea of how to fix this? the max_allowed_packet is also only set after a provision, the databases I configured however are getting loaded in but now doesn't work because of 1 field value being larger than the default 4 mb :(

**Comment by ethanf22** [comment]
I put a shell script in the `puphpet/files/startup-always` to with a mysql command `mysql -uroot -p123 -e "SET GLOBAL SQL_MODE='NO_ENGINE_SUBSTITUTION';" database_name` to fix my issue.

**Comment by pbowyer** [comment]
In testing, this issue is only present with MySQL. If I install MariaDB then `override_options` is applied on `vagrant up`. If I install MySQL then I have to do `vagrant up` followed by `vagrant provision`

Relevant extract (I flipped `install` for each test):
```
mariadb:
    install: '1'
    settings:
        version: '10.1'
        root_password: '123'
        override_options:
          'mysqld':
            'sql_mode': 'ONLY_FULL_GROUP_BY,STRICT_TRANS_TABLES,NO_ZERO_IN_DATE,ERROR_FOR_DIVISION_BY_ZERO,NO_AUTO_CREATE_USER,NO_ENGINE_SUBSTITUTION'
    users:
        mariadbnu_hy7p0pfo242g:
            name: dbuser
            password: '123'
    databases:
        mariadbnd_0arm4owr3vw6:
            name: dbname
            sql: ''
    grants:
        mariadbng_pnsujyxtl2ul:
            user: dbuser
            table: '*.*'
            privileges:
                - ALL
mysql:
    install: '0'
    settings:
        version: '5.7'
        root_password: testing123
        override_options:
          'mysqld':
            'sql_mode': 'ONLY_FULL_GROUP_BY,STRICT_TRANS_TABLES,NO_ZERO_IN_DATE,ERROR_FOR_DIVISION_BY_ZERO,NO_AUTO_CREATE_USER,NO_ENGINE_SUBSTITUTION'
    users:
        mysqlnu_5nentzb53ba2:
            name: dbuser
            password: '123'
    databases:
        mysqlnd_0btt41s0yqn7:
            name: dbname
            sql: /var/www/data/sql/database.sql
    grants:
        mysqlng_olmyupvvm05v:
            user: dbuser
            table: 'dbname.*'
            privileges:
                - ALL
```
Comparing https://github.com/puphpet/puppet-puphpet/blob/master/manifests/mysql/server.pp and https://github.com/puphpet/puppet-puphpet/blob/master/manifests/mariadb/server.pp#L12 the `override_options` does not look to be implemented for MySQL. This doesn't explain why `vagrant provision` will fix it on MySQL on its second run. I have tried modifying the code, no success yet.

**Comment by wallacio** [comment]
You're doing better that me if you can set sql_mode/sql-mode in the YAML. I don't see it being written anywhere during provisioning, not even ending up in a .bak file.

Using 5.7 param name of sql-mode:

```YAML
mysql:
    install: '1'
    settings:
        version: '5.7'
        root_password: '123'
        override_options:
          mysqld:
            sql-mode: 'NO_ZERO_IN_DATE,NO_ZERO_DATE,ERROR_FOR_DIVISION_BY_ZERO,NO_AUTO_CREATE_USER,NO_ENGINE_SUBSTITUTION'
```
I'm new to Puppet/hiera so I'm not quite sure what witchery occurs. Is it generally the case that in theory, any key/value added to `override_options::mysqld` will be added to my.cnf or does every possible option have to be present in the manifest?

**Comment by bastos71** [comment]
Where are we about this issue guys ?
